Mathlib/Tactic/Linter/UnusedInstancesInType.lean

```diff
@@ -0,0 +1,222 @@
+/-
+Copyright (c) 2025 Thomas R. Murrills. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas R. Murrills
+-/
+module
+
+public import Mathlib.Init
+public meta import Mathlib.Lean.Expr.Basic
+public meta import Mathlib.Lean.Environment
+public meta import Mathlib.Lean.Elab.InfoTree
+import Lean.Elab.Command
+
+/-!
+# Linters for Unused Instances in Types
+
+This file declares linters which detect certain instance hypotheses in declarations that are unused
+in the remainder of the type.
+
+Currently, these linters only handle theorems. (This also includes `lemma`s and `instance`s of
+`Prop` classes.)
+
+- `unusedDecidableInType` linter (currently off by default): suggests replacing type-unused
+  `Decidable*` instance hypotheses, and could therefore be replaced by `classical` in the proof.
+
+TODO: log on type signature instead of whole command
+TODO: add more linters!
+TODO: create Try This suggestions
+-/
+
+meta section
+
+open Lean Meta Elab Command Linter
+
+namespace Mathlib.Linter.UnusedInstancesInType
+
+/--
+A structure for storing information about a parameter of some declaration, usually within some
+telescope. We use this for recording the expressions and index associated to a parameter which was
+unused in the remainder of the type.
+
+`m!"{param}"` displays as `[{param.type?.get}] (#{param.idx + 1})` if `type?` is `some _`, and as
+`parameter #{param.idx + 1}` as a failsafe if `type?` is `none`. (We always expect `type?` to be
+`some _`, but would like a fallback if there are unexpected issues in telescoping.)
+-/
+private structure Parameter where
+  /- TODO: include syntax references to the binders here, and use the "real" fvars created during
+  elaboration if possible -/
+  /-- The free variable of the parameter created in a telescope for logging. -/
+  fvar? : Option Expr
+  /-- The type of the parameter created in a telescope for logging. -/
+  type? : Option Expr
```

Comment:
These should always be `some` in normal operation, right? Maybe let's say so here.